lib/domain.js

```diff
@@ -48,13 +48,61 @@ Object.defineProperty(process, 'domain', {
   }
 });
 
+let asyncHook = null;
+const pairing = new Map();
+function enableHook() {
+
+  if (asyncHook !== null) {
+    return;
+  }
+
+  asyncHook = async_hooks.createHook({
+    init(asyncId, type, triggerAsyncId, resource) {
+
+      if (process.domain !== null) { // if this operation is created while in a domain, let's mark it
+        pairing.set(asyncId, process.domain);
+        resource.domain = process.domain;
```

Comment:
Was the handle objects mutated before too?

Reply:
I don't thinks so, it was made on request https://github.com/nodejs/node/pull/16222#discussion_r144725755

Reply:
@addaleax could you explain why this is needed? I see no need to add extra `domain` complexity if it didn't already exist.

Reply:
@AndreasMadsen The thing is – it does already exist. With this, we can remove the code that sets it in other places, which would be very nice (look around for `domain_string()` in the source).

For event emitters (and therefore the handle wraps), the property is added in `events.js`, so we might need to come up with a better solution there.

Reply:
Oh, by the way; this is essentially how `MakeCallback` kept track of the domain – by looking at the resource object’s `.domain` property

Reply:
> The thing is – it does already exist.

Okay, then it is fine. I thought it was only `EventEmitter` and `Promise`.

Reply:
Yeah, those + all request types – that sums up everything in core …